README.md
# cras_imu_tools

Tools for working with IMU messages.

## `cras_imu_tools/gyro_bias_remover` Nodelet and `gyro_bias_remover` Node

Nodelet for measuring and removing gyro bias in IMU messages.

The nodelet expects that the robot (or at least its IMU) is stationary for some time, during which calibration samples are taken.

### Subscribed Topics

- `imu/data` (type `sensor_msgs/Imu`): The biased IMU data.
- `odom_cmd_vel` (type `nav_msgs/Odometry`): An odometry message telling the node whether the robot got some motion command. This can interrupt calibration.
- `reset` (any type): Whenever a message is received on this topic, the nodelet is reset, so for example calibration starts anew.

### Published Topics

- `imu_unbiased/data` (type `sensor_msgs/Imu`): The unbiased IMU data.
- `imu/gyro_bias` (type `geometry_msgs/Vector3Stamped`): The estimated gyro bias values.
- `emergency_stop` (type `std_msgs/Bool`): True is published to this topic during initial calibraiton. You can wire it to a motion disabling system of the robot.
- `emergency_stop/cmd_vel` (type `geometry_msgs/Twist`): Zero velocity commands are published to this topic during initial calibration. You can use it as a high priority motion command to disable motion of the robot.
- `diagnostics`: The bias remover can publish diagnostics if it is configured for it.
- `speak/info` (type `std_msgs/String`): Info-level messages from the calibration. If your robot has an audio output, route these messages to it so that you know the state of the calibration (e.g. via `sound_play` node).
- `speak/warn` (type `std_msgs/String`): Warning-level messages from the calibration. If your robot has an audio output, route these messages to it so that you know the state of the calibration (e.g. via `sound_play` node).
- `speak/err` (type `std_msgs/String`): Error-level messages from the calibration. If your robot has an audio output, route these messages to it so that you know the state of the calibration (e.g. via `sound_play` node).


### Parameters

- `~initial_calibration_duration` (float, default 15.0 s): Duration of the initial calibration procedure when the robot has to be stationary.
- `~min_calibration_samples` (uint, default 1000): The minimum number of IMU messages to collect during initial calibration.
- `~bias_estimation_rate` (float, default 0.01): Rate of change of the estimated bias. This is a parameter of a low-pass filter.
- `~gyro_not_moving_threshold` (float, default 0.05 rad/s): Values of gyro readings that can be treated as stationary. It is important that the bias is smaller than this value, otherwise the calibration will think that the robot was moving during sample collection.
- `~cmd_vel_threshold` (float, default 0.001 m/s or rad/s): Values of motion commands that can be treated as stationary.
- `~not_moving_duration_threshold` (float, default 2.0 s): After how many seconds online recalibration should be started if the robot was stationary for the whole time.
- `~skip_calibration` (bool, default false): If true, the initial calibration will not be performed and the node will right away start working with zero estimated bias.
- `~produce_diagnostics` (bool, default true): If true, the node will produce diagnostic messages.


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package cras_imu_tools
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2023-06-19)
------------------
* Removed cras imu transformer as it got fixed upstream
* Noetic compatibility.
* Publish commands to disable robot motion while doing the initial calibration.
* More fine-grained speak interface.
* Initial commit with gyro bias remover and fixed imu_transformer.
* Contributors: Martin Pecka


include/cras_imu_tools/transform_imu.h
#pragma once

/**
 * \file
 * \brief A corrected version of IMU transformer library which changes orientation correctly.
 * \author Paul Bovbel (+ changes from Martin Pecka)
 * SPDX-License-Identifier: BSD-3-Clause
 * SPDX-FileCopyrightText: Czech Technical University in Prague
 * 
 * This function in this file file has been copied from
 * https://github.com/ros-perception/imu_pipeline/blob/noetic-devel/imu_transformer/include/imu_transformer/tf2_sensor_msgs.h
 * and a change was made to resolve https://github.com/ros-perception/imu_pipeline/issues/8 .
 */

#include <Eigen/Core>

#include <geometry_msgs/TransformStamped.h>
#include <imu_transformer/tf2_sensor_msgs.h>
#include <sensor_msgs/Imu.h>

namespace cras
{

/**
* Transforms sensor_msgs::Imu data from one frame to another
*/
inline
void doTransform(const sensor_msgs::Imu &imu_in, sensor_msgs::Imu &imu_out, const geometry_msgs::TransformStamped& t_in)
{

  imu_out.header = t_in.header;

  // Discard translation, only use orientation for IMU transform
  Eigen::Quaternion<double> r(
    t_in.transform.rotation.w, t_in.transform.rotation.x, t_in.transform.rotation.y, t_in.transform.rotation.z);
  Eigen::Transform<double,3,Eigen::Affine> t(r);

  Eigen::Vector3d vel = t * Eigen::Vector3d(
    imu_in.angular_velocity.x, imu_in.angular_velocity.y, imu_in.angular_velocity.z);

  imu_out.angular_velocity.x = vel.x();
  imu_out.angular_velocity.y = vel.y();
  imu_out.angular_velocity.z = vel.z();

  tf2::transformCovariance(imu_in.angular_velocity_covariance, imu_out.angular_velocity_covariance, r);

  Eigen::Vector3d accel = t * Eigen::Vector3d(
    imu_in.linear_acceleration.x, imu_in.linear_acceleration.y, imu_in.linear_acceleration.z);


  imu_out.linear_acceleration.x = accel.x();
  imu_out.linear_acceleration.y = accel.y();
  imu_out.linear_acceleration.z = accel.z();

  tf2::transformCovariance(imu_in.linear_acceleration_covariance, imu_out.linear_acceleration_covariance, r);

  Eigen::Quaternion<double> orientation = Eigen::Quaternion<double>(
    imu_in.orientation.w, imu_in.orientation.x, imu_in.orientation.y, imu_in.orientation.z) * r.inverse();

  imu_out.orientation.w = orientation.w();
  imu_out.orientation.x = orientation.x();
  imu_out.orientation.y = orientation.y();
  imu_out.orientation.z = orientation.z();

  tf2::transformCovariance(imu_in.orientation_covariance, imu_out.orientation_covariance, r);

}

}

include/cras_imu_tools/gyro_bias_remover.h
#pragma once

/**
 * \file
 * \brief 
 * \author Martin Pecka
 * SPDX-License-Identifier: BSD-3-Clause
 * SPDX-FileCopyrightText: Czech Technical University in Prague
 */

#include <memory>

#include <geometry_msgs/Twist.h>
#include <geometry_msgs/Vector3Stamped.h>
#include <nav_msgs/Odometry.h>
#include <ros/console.h>
#include <ros/publisher.h>
#include <ros/subscriber.h>
#include <sensor_msgs/Imu.h>
#include <std_msgs/Bool.h>
#include <topic_tools/shape_shifter.h>

#include <cras_cpp_common/nodelet_utils.hpp>

namespace cras
{

enum class BiasObserverState
{
  INITIAL_CALIBRATION,
  MOVING,
  STOPPED_SHORT,
  STOPPED_LONG
};

class GyroBiasRemoverNodelet : public cras::Nodelet
{
protected:
  void onInit() override;
  
  void onImuMsg(const sensor_msgs::ImuConstPtr& msg);
  void onOdomMsg(const nav_msgs::OdometryConstPtr& msg);
  void estimateBias(const sensor_msgs::Imu& msg);

  void produceDiagnostics(diagnostic_updater::DiagnosticStatusWrapper& stat);
  void reportBiasChange();
  
  void speak(const std::string& message, ros::console::levels::Level level);
  
  void onReset(const topic_tools::ShapeShifter&);
  void reset();
  
private:
  std::unique_ptr<cras::DiagnosedPublisher<sensor_msgs::Imu>> imuPub;
  ros::Publisher biasPub;
  ros::Publisher speakInfoPub;
  ros::Publisher speakWarnPub;
  ros::Publisher speakErrPub;
  ros::Publisher stopCommandPub;
  ros::Publisher stopLockPub;
  ros::Subscriber imuSub;
  ros::Subscriber isMovingSub;
  ros::Subscriber resetSub;
  
  geometry_msgs::Vector3Stamped gyroBias;
  geometry_msgs::Twist stopCommand;
  std_msgs::Bool stopLockMsg;
  std_msgs::Bool stopUnlockMsg;
  
  BiasObserverState state {BiasObserverState::INITIAL_CALIBRATION};
  
  ros::Duration initialCalibrationDuration;
  size_t minCalibrationSamples {1000};
  double gyroBiasEstimationRate {0.01};
  bool skipCalibration {false};
  bool shouldProduceDiagnostics {true};
  
  double gyroThreshold {0};
  double cmdVelThreshold {0};
  ros::Duration notMovingDurationThreshold;
  
  ros::Duration notMovingDuration;
  ros::Time lastNotMovingTime;
  ros::Time lastReceiveTime;
  size_t numCalibrationSamples {0};
  ros::Time calibrationStartedTime;
  bool hasOdomMsg {false};
};

}

src/gyro_bias_remover.cpp
/**
 * \file
 * \brief 
 * \author Martin Pecka
 * SPDX-License-Identifier: BSD-3-Clause
 * SPDX-FileCopyrightText: Czech Technical University in Prague
 */

#include <geometry_msgs/Vector3Stamped.h>
#include <nodelet/nodelet.h>
#include <pluginlib/class_list_macros.h>
#include <topic_tools/shape_shifter.h>
#include <std_msgs/String.h>

#include <cras_cpp_common/type_utils/literal_sz.h>

#include <cras_imu_tools/gyro_bias_remover.h>

namespace cras
{

void GyroBiasRemoverNodelet::onInit()
{
  auto nh = this->getNodeHandle();
  auto pnh = this->getPrivateNodeHandle();
  auto params = this->privateParams();
  
  this->initialCalibrationDuration = params->getParam("initial_calibration_duration", ros::Duration(15), "s");
  this->minCalibrationSamples = params->getParam("min_calibration_samples", 1000_sz, "IMU samples");
  this->gyroBiasEstimationRate = params->getParam("bias_estimation_rate", 0.01);
  this->gyroThreshold = params->getParam("gyro_not_moving_threshold", 0.05, "rad/s");
  this->cmdVelThreshold = params->getParam("cmd_vel_threshold", 1e-3, "m/s or rad/s");
  this->notMovingDurationThreshold = params->getParam("not_moving_duration_threshold", ros::Duration(2), "s");
  this->skipCalibration = params->getParam("skip_calibration", false);
  this->shouldProduceDiagnostics = params->getParam("produce_diagnostics", true);
  
  if (this->skipCalibration)
    this->state = BiasObserverState::MOVING;
  
  this->calibrationStartedTime = ros::Time::now();
  this->stopLockMsg.data = true;
  this->stopUnlockMsg.data = false;
  
  this->imuPub = this->advertiseDiagnosed<sensor_msgs::Imu>(nh, pnh, "output", "imu_unbiased/data", 100);
  this->biasPub = nh.advertise<geometry_msgs::Vector3Stamped>("imu/gyro_bias", 100, true);
  this->speakInfoPub = nh.advertise<std_msgs::String>("speak/info", 2);
  this->speakWarnPub = nh.advertise<std_msgs::String>("speak/warn", 2);
  this->speakErrPub = nh.advertise<std_msgs::String>("speak/err", 2);
  this->stopCommandPub = nh.advertise<geometry_msgs::Twist>("emergency_stop/cmd_vel", 2);
  this->stopLockPub = nh.advertise<std_msgs::Bool>("emergency_stop", 2);

  this->imuSub = nh.subscribe("imu/data", 100, &GyroBiasRemoverNodelet::onImuMsg, this);
  this->isMovingSub = nh.subscribe("odom_cmd_vel", 100, &GyroBiasRemoverNodelet::onOdomMsg, this);
  this->resetSub = nh.subscribe("reset", 1, &GyroBiasRemoverNodelet::onReset, this);
  
  if (this->shouldProduceDiagnostics)
  {
    this->getDiagUpdater().add("status", this, &GyroBiasRemoverNodelet::produceDiagnostics);
    this->startDiagTimer();
  }
  
  CRAS_INFO("IMU calibration has started.");
  this->speak("Calibrating gyro, do not move me.", ros::console::levels::Warn);
}

void GyroBiasRemoverNodelet::onImuMsg(const sensor_msgs::ImuConstPtr& msg)
{
  if (ros::Time::now() + ros::Duration(3) < this->lastReceiveTime)
  {
    CRAS_WARN("ROS time has jumped back, resetting.");
    this->reset();
  }
  this->lastReceiveTime = ros::Time::now();
  
  this->estimateBias(*msg);

  if (this->state == BiasObserverState::INITIAL_CALIBRATION)
    return;
  
  sensor_msgs::Imu unbiased = *msg;
  unbiased.angular_velocity.x -= this->gyroBias.vector.x;
  unbiased.angular_velocity.y -= this->gyroBias.vector.y;
  unbiased.angular_velocity.z -= this->gyroBias.vector.z;

  this->imuPub->publish(unbiased);
}

void GyroBiasRemoverNodelet::onOdomMsg(const nav_msgs::OdometryConstPtr& msg)
{
  this->hasOdomMsg = true;
  CRAS_DEBUG_THROTTLE(1.0, "State is %i", static_cast<std::underlying_type<BiasObserverState>::type>(this->state));
  
  const auto& v = msg->twist.twist.linear;
  const auto& w = msg->twist.twist.angular;
  
  const auto& th = this->cmdVelThreshold;
  
  const auto isNowMoving = std::abs(v.x) > th || std::abs(v.y) > th || std::abs(v.z) > th || std::abs(w.x) > th
    || std::abs(w.y) > th || std::abs(w.z) > th;
  
  if (isNowMoving && this->state == BiasObserverState::INITIAL_CALIBRATION)
  {
    CRAS_ERROR("Robot has moved during IMU calibration!");
    this->speak("Gyro calibration failed, I moved!", ros::console::levels::Error);
  }
  
  if (isNowMoving)
  {
    if (this->state != BiasObserverState::INITIAL_CALIBRATION)
    {
      if (this->state == BiasObserverState::STOPPED_LONG)
        this->reportBiasChange();
      this->state = BiasObserverState::MOVING;
    }
    this->notMovingDuration = {0, 0};
    this->lastNotMovingTime = {0, 0};
    return;
  }
  
  if (this->state == BiasObserverState::MOVING)
  {
    this->state = BiasObserverState::STOPPED_SHORT;
    this->lastNotMovingTime = msg->header.stamp;
    this->notMovingDuration = {0, 0};
    return;
  }

  this->notMovingDuration += msg->header.stamp - this->lastNotMovingTime;
  this->lastNotMovingTime = msg->header.stamp;
  if (this->state != BiasObserverState::INITIAL_CALIBRATION)
  {
    if (this->notMovingDuration > this->notMovingDurationThreshold)
      this->state = BiasObserverState::STOPPED_LONG;
    else
      this->state = BiasObserverState::STOPPED_SHORT;
  }
}

void GyroBiasRemoverNodelet::estimateBias(const sensor_msgs::Imu& msg)
{
  if (this->state != BiasObserverState::INITIAL_CALIBRATION && this->state != BiasObserverState::STOPPED_LONG)
    return;

  const auto& w = msg.angular_velocity;
  auto& bias = this->gyroBias.vector;
  
  if (this->state == BiasObserverState::INITIAL_CALIBRATION)
  {
    bias.x += w.x;
    bias.y += w.y;
    bias.z += w.z;
    this->numCalibrationSamples++;

    const auto calibDuration = msg.header.stamp - this->calibrationStartedTime;
    if (this->numCalibrationSamples >= this->minCalibrationSamples && calibDuration >= this->initialCalibrationDuration)
    {
      bias.x /= static_cast<double>(this->numCalibrationSamples);
      bias.y /= static_cast<double>(this->numCalibrationSamples);
      bias.z /= static_cast<double>(this->numCalibrationSamples);
      
      this->gyroBias.header = msg.header;
      this->biasPub.publish(this->gyroBias);
      
      if (this->lastNotMovingTime == ros::Time(0, 0))
        this->state = BiasObserverState::MOVING;
      else if (this->notMovingDuration >= this->notMovingDurationThreshold)
        this->state = BiasObserverState::STOPPED_LONG;
      else
        this->state = BiasObserverState::STOPPED_SHORT;
      CRAS_WARN("IMU calibration finished.");
      this->speak("Gyros calibrated!", ros::console::levels::Warn);
      this->stopLockPub.publish(this->stopUnlockMsg);
      this->reportBiasChange();
    }
    else
    {
      CRAS_WARN_THROTTLE(1.0, "IMU is calibrating, do not move the robot.");
      this->stopCommandPub.publish(this->stopCommand);
      this->stopLockPub.publish(this->stopLockMsg);
    }
  }
  else
  {
    auto th = this->gyroThreshold;

    if (std::abs(w.x - bias.x) > th || std::abs(w.y - bias.y) > th || std::abs(w.z - bias.z) > th)
    {
      this->state = BiasObserverState::MOVING;
      this->lastNotMovingTime = {0, 0};
      this->notMovingDuration = {0, 0};
      this->reportBiasChange();
      return;
    }

    const auto& rate = this->gyroBiasEstimationRate;

    bias.x = (1 - rate) * bias.x + rate * w.x;
    bias.y = (1 - rate) * bias.y + rate * w.y;
    bias.z = (1 - rate) * bias.z + rate * w.z;

    this->gyroBias.header.frame_id = msg.header.frame_id;
    this->gyroBias.header.stamp = msg.header.stamp;

    this->biasPub.publish(this->gyroBias);
  }
}

void GyroBiasRemoverNodelet::reset()
{
  this->state = this->skipCalibration ? BiasObserverState::MOVING : BiasObserverState::INITIAL_CALIBRATION;
  this->numCalibrationSamples = 0;
  this->gyroBias.vector.x = this->gyroBias.vector.y = this->gyroBias.vector.z = 0;
  this->notMovingDuration = {0, 0};
  this->lastNotMovingTime = {0, 0};
  this->calibrationStartedTime = ros::Time::now();
  this->hasOdomMsg = false;
  if (this->shouldProduceDiagnostics)
    this->getDiagUpdater().force_update();
}

void GyroBiasRemoverNodelet::onReset(const topic_tools::ShapeShifter&)
{
  this->reset();
}

void GyroBiasRemoverNodelet::produceDiagnostics(diagnostic_updater::DiagnosticStatusWrapper& stat)
{
  if (this->state == BiasObserverState::INITIAL_CALIBRATION)
    stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "Initial calibration");
  else if (this->hasOdomMsg)
    stat.summary(diagnostic_msgs::DiagnosticStatus::OK, "Running");
  else
    stat.summary(diagnostic_msgs::DiagnosticStatus::WARN, "No odom messages received");

  switch (this->state)
  {
    case BiasObserverState::INITIAL_CALIBRATION:
      stat.add("State", "Initial calibration");
      stat.addf("Num samples", "%lu", this->numCalibrationSamples);
      stat.add("Duration of calibration", cras::to_string(ros::Time::now() - this->calibrationStartedTime));
      break;
    case BiasObserverState::MOVING:
      stat.add("State", "Moving");
      break;
    case BiasObserverState::STOPPED_SHORT:
      stat.add("State", "Just stopped");
      stat.add("Last motion before", cras::to_string(this->notMovingDuration));
      break;
    case BiasObserverState::STOPPED_LONG:
      stat.add("State", "Standing still, calibrating bias");
      stat.add("Last motion before", cras::to_string(this->notMovingDuration));
      break;
  }
}

void GyroBiasRemoverNodelet::reportBiasChange()
{
  CRAS_INFO("Estimated gyro bias is: x=%.6f y=%.6f z=%.6f",
    this->gyroBias.vector.x, this->gyroBias.vector.y, this->gyroBias.vector.z);
}

void GyroBiasRemoverNodelet::speak(const std::string& message, const ros::console::levels::Level level)
{
  std_msgs::String msg;
  msg.data = message;
  
  switch (level)
  {
    case ros::console::levels::Info:
      this->speakInfoPub.publish(msg);
      break;
    case ros::console::levels::Warn:
      this->speakWarnPub.publish(msg);
      break;
    default:
      this->speakErrPub.publish(msg);
      break;
  }
}

}

PLUGINLIB_EXPORT_CLASS(cras::GyroBiasRemoverNodelet, nodelet::Nodelet)